lib/downloadhelper.cpp

```diff
@@ -14,29 +15,46 @@
 
 #include <QFileInfo>
 
+constexpr int FILEEXTENSION_MAX_LENGTH = 32;
+constexpr int FILENAME_MAX_LENGTH = 255;
+
 SailfishOS::WebEngineUtils::DownloadHelper::DownloadHelper(QObject *parent)
     : QObject(parent)
 {
 }
 
-QString SailfishOS::WebEngineUtils::DownloadHelper::createUniqueFileUrl(const QString &fileName, const QString &path) const
+QString SailfishOS::WebEngineUtils::DownloadHelper::createUniqueFileUrl(QString fileName, const QString &path) const
 {
     if (path.isEmpty() || fileName.isEmpty()) {
         return QString();
     }
 
-    const QFileInfo fileInfo(fileName);
-    const QString newFile("%1/%2(%3)%4%5");
-    const QString baseName = fileInfo.baseName();
-    const QString suffix = fileInfo.completeSuffix();
+    // Replace illegal characters with dashes
+    fileName.replace(QRegExp("[<>:\"|?*%/\\\\-]+"), "-");
 
-    QString result(path + "/" + fileName);
-    int collisionCount(1);
+    // Replace whitespace characters with underscores
+    fileName.replace(QRegExp("[\\s_-]+"), "_");
 
-    while (QFileInfo::exists(result)) {
-        collisionCount++;
-        result = newFile.arg(path).arg(baseName).arg(collisionCount).arg(suffix.isEmpty() ? "" : ".").arg(suffix);
+    // Determine of extension position
+    int dotPosition = fileName.length();
```

Comment:
lastIndexOf('.')

Reply:
Actually even better is to use QFileInfo and completeSuffix so that combined extensions are covered as well .tag.gz like cases.

Reply:
If pass a file name of, for example, 300 characters, with a leading dot, completeSuffix will return it in its entirety. This is incorrect, since there are no such long extensions. And this does not allow shortening the filename to 255 symbols, since only basename can be shortened, which is empty in this case. Limiting the length of the extension to some reasonable value solves this problem.

Reply:
Leading dot and your case is not resulting right extension either -- it's just something that is fabricated.

Not true either with one leading dot suffix is empty. When there are n leading dots then completeSuffix is n-1 dots. This can be consider valid as well. But if you think that leading dots are not good... then something like this would do it (wrote without compiling):

```
while (fileName.endsWith('.') && fileName.length() > 0) {
    fileName.chop(1);
}

if (fileName.isEmpty() {
    fileName = QStringLiteral("unknown_file"); // or something like
}
const QFileInfo fileInfo(fileName);
QString baseName = fileInfo.baseName();
const QString suffix = fileInfo.completeSuffix();

int maxLen = 255 - suffix.length();
baseName = baseName.left(maxLen);
```

Reply:
Look into QMimeDatabase. It has a suffixForFileName() function which should return the correct suffix for known extensions, and if the mime type is known but there's no extension in the file name it should be able to provide one.